contracts/EventBasedPredictionMarket.sol

```diff
@@ -0,0 +1,333 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/utils/math/Math.sol";
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import "@uma/core/contracts/common/implementation/ExpandedERC20.sol";
+import "@uma/core/contracts/common/implementation/FixedPoint.sol";
+import "@uma/core/contracts/common/implementation/Lockable.sol";
+import "@uma/core/contracts/common/implementation/Testable.sol";
+import "@uma/core/contracts/oracle/implementation/Constants.sol";
+
+import "@uma/core/contracts/common/interfaces/ExpandedIERC20.sol";
+import "@uma/core/contracts/oracle/interfaces/FinderInterface.sol";
+import "@uma/core/contracts/oracle/interfaces/OptimisticOracleInterface.sol";
+
+import "@uma/core/contracts/financial-templates/common/financial-product-libraries/long-short-pair-libraries/BinaryOptionLongShortPairFinancialProductLibrary.sol";
+
+// TODO use OptimisticOracleInterface from @uma/core once it's updated with setEventBased
+interface OptimisticOracleInterfaceEventBased {
+    function setEventBased(
+        bytes32 identifier,
+        uint256 timestamp,
+        bytes memory ancillaryData
+    ) external;
+}
+
+contract EventBasedPredictionMarket is Testable, Lockable {
+    using FixedPoint for FixedPoint.Unsigned;
+    using SafeERC20 for IERC20;
+
+    /*************************************
+     *  EVENT BASED PREDICTION MARKET DATA STRUCTURES  *
+     *************************************/
+
+    // Define the contract's constructor parameters as a struct to enable more variables to be specified.
+    struct ConstructorParams {
```

Comment:
I dont think this needs to be a struct any more now that you've drastically decreased the number of props in the constructor. this was originally done like this to accommodate the limitation in constructor param lengths.